mdn/developer-portal
# 139: Fix auto-creation of static build following publish/unpublish

This changeset ensures that an automatic build-and-sync-to-S3 of the site takes place when a page is published or unpublished.

This approach uses a task queue (Celery) rather than spawning a new `subprocess.Process`. This should be more stable and visible.

(Resolves #139)

## Key changes:

* Update Dockerfile to:
  - rename `app` target as `app_base`, to make it clearer it's being used for multiple things
  - not start gunicorn as part of `app_base`'s build, because that's not needed and is handled by docker-compose or k8s

* Update docker-compose.yml to:
  - add `redis` container
  - update the `app` config to inherit from a new `common` config, which uses the `app_base` target
  - add new `worker` container, inheriting from the `common` config, too

* Add `celery` to the project
  - Stop the use of `subprocess.Process` when a page is published or unpublished.
  - Use a a properly enqueued task to handle the background dump-to-static-then-upload-to-S3 process

*  Add `django-celery-results` to give us easier visibility of job success or failure

    This is a temporary and basic solution to give us an eye on whether background tasks are
happy, but it is limited because it only displays tasks after they have completed (ie
once they have a result).
    Switching to something that lets us see in-flight tasks would be useful, down the line. (eg [Flower](https://flower.readthedocs.io/en/latest/features.html), though that's a sledgehammer to crack a nut)

*  Ensure Django (and therefore the Celery worker) can get hold of the credentials Celery needs

  - Update Makefile to export AWS-related config to the environment
  - In Django, pick up AWS-related config from environment

    This change assumes the AWS-related config is already available at in the general config
at the k8s level -- this seems reasonable, because they are all already mentioned in the
k8s config files.

*  Add light integration tests to show that publishing and unpublishing will trigger static-build-and-sync behaviour

## Things to note

* The standard pipeline of `manage.py build` and `manage.py publish` results in **ALL CMS pages always being uploaded, even if they are not directly affected by any changes**. However, only changed staticfiles are uploaded each time. For now, the 'all the pages' approach should be OK, but I expect we'll need to revisit that once the amount of content increases.

* We need to do a bit of thinking about how this background publishing approach will feel in reality:
    * Related to https://github.com/mdn/developer-portal/issues/252, the UI currently makes no mention of the static site, and therefore no mention that the static site will take X seconds to be baked and synced to S3. This might be something we can solve with documentation/editor education, but it would be good to come up with a more robust solution
   * django-celery-results shows us when build-and-sync has completed, but not when it's in progress - Flower should help there
   * We do need to enable hourly scheduled publishing, so that things like events listings naturally decay/update over the course of a day, rather than waiting for the next time an editor publishes content. However, there's a (minor?) risk of race conditions causing confusing output if - essentially - two separate tasks are publishing the site at the same time. If this is a problem, we can look to get around it by checking what's in Celery's Redis before deciding whether to do a build-and-sync, or not. Up for debate, for sure!

* After running into a speedbump with Celery, I tried `dramatiq` based on @escattone's suggestion. However, `dramatiq` automatically tries to import any module called `tasks.py`, which causes `django-bakery` to explode, because django-bakery [_expects_ Celery to be available when its tasks module is imported](https://github.com/datadesk/django-bakery/blob/master/bakery/tasks.py#L6)...

* This could definitely do with some laser-eyes from @escattone and @limed re how it'll slot into our staging and production infra. I have opened a soft-blocking issue (https://github.com/mdn/developer-portal/issues/255) to add Redis, because Celery needs it.

* If you want to run the async build locally, you'll need to set the following in your `settings/local.py`:
    * `AWS_ACCESS_KEY_ID`
    * `AWS_SECRET_ACCESS_KEY`
    * `AWS_BUCKET_NAME`
    * `AWS_REGION`
and set `DEBUG = False`

* Note, too that the Celery `worker` will need manually restarting to pick up code changes. It doesn't auto-detect them like `runserver` does. 



## How to test

- check out this branch
- run migrations (for `django-celery-results`' tables)
- add/update your `settings.local` to include the four AWS settings mentioned above, as well as turning DEBUG mode off. (Make sure you don't use any production-type AWS credentials for this testing!!) 
- ensure the target S3 bucket exists
- boot up the stack with `docker-compose build && docker-compose up`
   - if you get problems, going nuclear with `docker-compose down --volumes && docker-compose build && docker-compose up` and/or using the `dev-setup --purge` should rebuild all you need. You may need to create a new Django/Wagtail superuser to log in to the CMS
- Pick a page and publish it.
- Check the console to see `worker_1` baking a static site then syncing it to S3
- Check the S3 bucket to see the site contents all there. Download the bucket and view in a local browser if you want.
- Remove the baked site from your S3 bucket
